### PR TITLE
[ROCm] Fix invalid max_seqlen in test_dummy_mha_with_nt, unskip NT tests

### LINKED ISSUE #168784: TestModule: test/test_nestedtensor.py::TestNestedTensorSubclass
## Context
- Source file: `test/test_nestedtensor.py`
- Test class: `TestNestedTensorSubclass`
- ROCm-skipped tests: 13
- Code reference: https://github.com/pytorch/pytorch/blob/main/test/test_nestedtensor.py

cc @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @jataylo @hongxiayang @naromero77amd @pragupta @jerrymannil @xinyazhang @dllehr-amd

### LINKED ISSUE #168791: Test: TestNestedTensorSubclass.test_dummy_mha_with_nt
## Context
- Source file: `test/test_nestedtensor.py`
- Test class: `TestNestedTensorSubclass`
- Test name: `test_dummy_mha_with_nt`
- Defined at line: 7388
- Parent issue: #168784
- Code reference: https://github.com/pytorch/pytorch/blob/main/test/test_nestedtensor.py

cc @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @jataylo @hongxiayang @naromero77amd @pragupta @jerrymannil @xinyazhang @dllehr-amd

### PR DESCRIPTION
**Pull request by geozhai**
FIXES #168784
FIXES #168791

### Summary

`test_dummy_mha_with_nt` declared cached `max_seqlen` values that were smaller than the true maximum sequence length. An understated cached value flows straight into the varlen SDPA kernel launch bound, so attention is silently truncated and gradients come out wrong. This is invalid test metadata rather than an SDPA backend problem, and nothing about it is ROCm-specific -- the same latent flakiness is reachable on NVIDIA. It surfaced on ROCm, which is why the test carried `@skipCUDAIfRocm`.

### Reproducing

Remove `@skipCUDAIfRocm` from `test_dummy_mha_with_nt` and run repeatedly against a shared Inductor cache:

```
for i in 1 2 3 4 5; do
  TORCHINDUCTOR_CACHE_DIR=/tmp/indcache \
    python test/test_nestedtensor.py -k test_dummy_mha_with_nt
done
```

Run 1 passes on a cold cache; runs 2 onward fail on `use_legacy_api_True_cuda` at the compiled-vs-eager gradient comparison. A fresh cache directory every run passes every time, which is why the failure looks flaky.

Cache logging shows the split:

```
Run 1: LocalAOTAutogradCache {hit: 0, miss: 2}  LocalFxGraphCache {hit: 0, miss: 5}  PASSED
Run 2: LocalAOTAutogradCache {hit: 2, miss: 0}  LocalFxGraphCache {hit: 5, miss: 0}  FAILED
```

with the cache-hit gradient wrong in one row:

```
max_abs_diff: 0.9519949555397034
mismatched_elements: 4 / 120
bad_rows: [24]
```

### Root cause

`offsets = torch.tensor([0, 2, 3, 30])` describes sequence lengths `[2, 1, 27]`, so the true maximum is 27. The test declared `max_length_1 = 10` for key and `max_length_2 = 20` for value, both understating it. `_cumulative_and_max_seq_len_nnz` reads the cached value and hands it to the flash / mem-efficient kernels as the launch bound, so the tail of the 27-long sequence is never attended to.

Two things this does not fully explain, stated plainly rather than glossed. The kv-side bound comes from `key`, not `value` -- `torch/nested/_internal/sdpa.py:629-633` reads `k_t` and `v_t`'s cached max is never queried on this path -- so the operative bound was 10, and a single corrupt row at 24 is not the signature of truncating at 10 or 20. And the same declared metadata produces correct gradients on a cache miss and wrong ones on a cache hit, even though `max_seqlen` is a Python int baked into the traced graph both times; that divergence is device-independent, and making the metadata valid hides it rather than fixing it. The exact mechanism linking the invalid bound to the observed row is unidentified; the fix is a metadata-validity correction whose effect on the symptom is empirical.

### Changes

Declare `max_seqlen_key = 27` and `max_seqlen_value = 28`. Both are at or above the true maximum so neither truncates, and both stay below `total_length = 30`, which `flash_attn` backward requires. They are kept distinct deliberately: the two assertions at the end of the test check that key and value carry independent metadata caches and that a declared value survives fx-trace plus `torch.compile`. Collapsing both to 27 would make them vacuous, since 27 is exactly `offsets.diff().max()` and `_get_max_seqlen()` recomputes that lazily when no cache is present.

Also removes `@skipCUDAIfRocm` from `test_dummy_mha_with_nt`, removes `@skipIfRocm` from `test_index_put_error` (its issue #175482 is closed, and the test passes -- the out-of-bounds `index_put` terminates the subprocess non-zero, which is all it asserts), restores the `total_length > max_seqlen` comment that documents the flash-backward constraint, and fixes a SPDA -> SDPA typo. With both skips gone, `TestNestedTensorSubclass` has no ROCm skips left, so `FIXES #168784` is accurate.

### Validation

Both tests were run on all four ROCm CI architectures -- gfx90a, gfx942, gfx950 and gfx1100 -- passing everywhere (`test_index_put_error` 2 tests OK, `test_dummy_mha_with_nt` 4 tests OK with the 2 CPU variants skipped per #133644). The pinned-cache reproduction above was confirmed on gfx90a and gfx950: unfixed, run 1 passes and runs 2-5 fail; fixed, all five pass.

One behavioural difference worth recording: `test_index_put_error` terminates via SIGABRT on gfx90a and gfx942 (device-side assert, HSA queue exception) but via a Python-level `AcceleratorError` exiting 1 on gfx1100. The test only asserts a non-zero exit, so both satisfy it.

Portions of this PR were prepared with an AI assistant.

cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @jataylo @hongxiayang @naromero77amd @pragupta @jerrymannil @xinyazhang

### CODE REVIEW

**Comment by jeffdaily** [comment]
Pushed a merge of latest viable/strict plus a maintainer commit, and updated the title and description.

The merge is needed regardless of review: this base is 609 commits back, far enough that the CI docker image for its `.ci/docker` hash has been pruned, so jobs die in `Initialize containers` with `ErrImagePull` and restarting never helps. It also brings in #192122.

The maintainer commit keeps key and value at distinct declared values (27 and 28) rather than collapsing both to 27. With a single value equal to `offsets.diff().max()`, the two cache assertions at the end of the test pass even if the compiled path discards the declared cache, and key/value become metadata-identical so aliasing between their caches is undetectable. It also drops the `@skipIfRocm` on `test_index_put_error`, which was the last ROCm skip in the class and the reason `FIXES #168784` was not accurate as written.

Validated both tests on gfx90a, gfx942, gfx950 and gfx1100. Your failure reproduces on gfx90a and gfx950 once `TORCHINDUCTOR_CACHE_DIR` is pinned across runs, and goes away with the fix -- I folded that into the description since it is stronger evidence than the flake-finder run.

I also softened the root-cause section: the kv bound comes from `key`, not `value`, so the operative value was 10 and a single bad row at 24 does not follow from truncating at either bound, and the cache-hit vs cache-miss divergence is device-independent and not fixed here. The description now says the mechanism is unidentified and the fix is empirical, which I think is defensible as-is.

Analysis assisted by AI.

**Review by jeffdaily — APPROVED** [review]
